File: src-tauri/src/accounts.rs
//! Account model, `accounts.json` persistence, and keyring secret helpers.
//!
//! Non-secret account configuration is stored in
//! `$XDG_CONFIG_HOME/cosmic-mail/accounts.json`. Secrets (IMAP passwords, Gmail
//! OAuth refresh tokens) live only in the Secret Service via the `keyring`
//! crate and are never written to disk or logged.

use std::io::Write;
use std::path::PathBuf;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

/// Keyring service name (shared by all secrets for this app).
pub const KEYRING_SERVICE: &str = "dev.pluscosmic.mail";

/// The kind of account, determining the auth mechanism used for IMAP/SMTP.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AccountKind {
    /// Plain IMAP with a password stored in the keyring.
    Imap,
    /// Gmail via OAuth2/XOAUTH2.
    Gmail,
}

/// Full account record persisted to `accounts.json` (contains no secrets).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Account {
    /// UUID v4 stable identifier.
    pub id: String,
    /// Primary email address.
    pub email: String,
    /// Human display name.
    pub display_name: String,
    /// Account kind (auth mechanism).
    pub kind: AccountKind,

    // Connection details. For Gmail these are hardcoded to the Gmail servers.
    /// IMAP host, e.g. `imap.gmail.com`.
    pub imap_host: String,
    /// IMAP port (implicit TLS, usually 993).
    pub imap_port: u16,
    /// SMTP host, e.g. `smtp.gmail.com`.
    pub smtp_host: String,
    /// SMTP port (STARTTLS 587 or implicit 465).
    pub smtp_port: u16,
    /// Login username (often equal to `email`).
    pub username: String,
}

/// Public projection of [`Account`] returned to the frontend (no connection
/// details or secrets — only what the UI needs).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AccountPublic {
    pub id: String,
    pub email: String,
    pub display_name: String,
    pub kind: AccountKind,
}

impl Account {
    /// Project to the public wire type.
    pub fn public(&self) -> AccountPublic {
        AccountPublic {
            id: self.id.clone(),
            email: self.email.clone(),
            display_name: self.display_name.clone(),
            kind: self.kind,
        }
    }
}

/// Path to `accounts.json`: `$XDG_CONFIG_HOME/cosmic-mail/accounts.json`.
pub fn accounts_path() -> Result<PathBuf> {
    let dir = dirs::config_dir()
        .context("could not determine config dir (XDG_CONFIG_HOME)")?
        .join("cosmic-mail");
    Ok(dir.join("accounts.json"))
}

/// Load all accounts from disk, returning an empty list if the file is absent.
pub fn load_accounts() -> Result<Vec<Account>> {
    let path = accounts_path()?;
    if !path.exists() {
        return Ok(Vec::new());
    }
    let data = std::fs::read(&path).with_context(|| format!("reading {}", path.display()))?;
    let accounts: Vec<Account> = serde_json::from_slice(&data).context("parsing accounts.json")?;
    Ok(accounts)
}

/// Atomically persist the account list to `accounts.json`.
///
/// Writes to a sibling temp file and renames into place so a crash mid-write
/// cannot corrupt the config.
pub fn save_accounts(accounts: &[Account]) -> Result<()> {
    let path = accounts_path()?;
    let dir = path
        .parent()
        .context("accounts.json has no parent directory")?;
    std::fs::create_dir_all(dir).with_context(|| format!("creating {}", dir.display()))?;

    let json = serde_json::to_vec_pretty(accounts).context("serializing accounts")?;

    let mut tmp = tempfile::NamedTempFile::new_in(dir)
        .with_context(|| format!("creating temp file in {}", dir.display()))?;
    tmp.write_all(&json).context("writing accounts temp file")?;
    tmp.flush().context("flushing accounts temp file")?;
    tmp.persist(&path)
        .with_context(|| format!("persisting {}", path.display()))?;
    Ok(())
}

// --- keyring helpers ---------------------------------------------------------

fn imap_password_key(account_id: &str) -> String {
    format!("imap-password:{account_id}")
}

fn oauth_refresh_token_key(account_id: &str) -> String {
    format!("oauth-refresh-token:{account_id}")
}

fn entry(user_key: &str) -> Result<keyring::Entry> {
    keyring::Entry::new(KEYRING_SERVICE, user_key)
        .with_context(|| format!("opening keyring entry {user_key}"))
}

/// Store an IMAP password for the given account.
pub fn set_imap_password(account_id: &str, password: &str) -> Result<()> {
    entry(&imap_password_key(account_id))?
        .set_password(password)
        .context("storing IMAP password in keyring")?;
    Ok(())
}

/// Fetch the IMAP password for the given account.
pub fn get_imap_password(account_id: &str) -> Result<String> {
    entry(&imap_password_key(account_id))?
        .get_password()
        .context("reading IMAP password from keyring")
}

/// Store the Gmail OAuth refresh token for the given account.
pub fn set_oauth_refresh_token(account_id: &str, token: &str) -> Result<()> {
    entry(&oauth_refresh_token_key(account_id))?
        .set_password(token)
        .context("storing OAuth refresh token in keyring")?;
    Ok(())
}

/// Fetch the Gmail OAuth refresh token for the given account.
pub fn get_oauth_refresh_token(account_id: &str) -> Result<String> {
    entry(&oauth_refresh_token_key(account_id))?
        .get_password()
        .context("reading OAuth refresh token from keyring")
}

/// Best-effort deletion of all secrets for an account (ignores missing entries).
pub fn delete_secrets(account_id: &str, kind: AccountKind) {
    let key = match kind {
        AccountKind::Imap => imap_password_key(account_id),
        AccountKind::Gmail => oauth_refresh_token_key(account_id),
    };
    if let Ok(entry) = entry(&key) {
        // Ignore NoEntry / other deletion errors — this is best effort.
        let _ = entry.delete_credential();
    }
}

/// Remove an account from the on-disk list, if present.
pub fn remove_account_from_disk(account_id: &str) -> Result<Option<Account>> {
    let mut accounts = load_accounts()?;
    if let Some(pos) = accounts.iter().position(|a| a.id == account_id) {
        let removed = accounts.remove(pos);
        save_accounts(&accounts)?;
        Ok(Some(removed))
    } else {
        Ok(None)
    }
}

/// Append an account to the on-disk list and persist.
pub fn add_account_to_disk(account: &Account) -> Result<()> {
    let mut accounts = load_accounts()?;
    accounts.push(account.clone());
    save_accounts(&accounts)?;
    Ok(())
}

File: internal/accounts/accounts.go
// Package accounts holds the account model, `accounts.json` persistence, and
// the keyring secret helpers.
//
// Non-secret account configuration is stored in
// `$XDG_CONFIG_HOME/cosmic-mail/accounts.json`. Secrets (IMAP passwords, Gmail
// OAuth refresh tokens) live only in the Secret Service keyring and are never
// written to disk or logged.
package accounts

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"

	"github.com/zalando/go-keyring"

	"cosmicmail/internal/buildinfo"
	"cosmicmail/internal/models"
	"cosmicmail/internal/settings"
	"cosmicmail/internal/xdg"
)

// KeyringService is the Secret Service `service` attribute shared by every
// secret this app stores. It matches the Rust build (keyring v4 also used the
// `service` + `username` attributes), so existing entries stay readable.
const KeyringService = "dev.pluscosmic.mail"

// Account is the full record persisted to accounts.json (contains no secrets).
type Account struct {
	ID          string             `json:"id"`
	Email       string             `json:"email"`
	DisplayName string             `json:"displayName"`
	Kind        models.AccountKind `json:"kind"`
	// Connection details. For Gmail these are the fixed Gmail servers.
	ImapHost string `json:"imapHost"`
	ImapPort int    `json:"imapPort"`
	SmtpHost string `json:"smtpHost"`
	SmtpPort int    `json:"smtpPort"`
	Username string `json:"username"`
}

// Public projects the account onto the wire type (no connection details).
func (a Account) Public() models.Account {
	return models.Account{ID: a.ID, Email: a.Email, DisplayName: a.DisplayName, Kind: a.Kind}
}

// Path is `$XDG_CONFIG_HOME/cosmic-mail/accounts.json`.
func Path() (string, error) {
	dir, err := xdg.AppConfigDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "accounts.json"), nil
}

// Load reads every account from disk; a missing file is an empty list.
func Load() ([]Account, error) {
	path, err := Path()
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(path)
	if errors.Is(err, os.ErrNotExist) {
		return []Account{}, nil
	}
	if err != nil {
		return nil, fmt.Errorf("reading %s: %w", path, err)
	}
	var accounts []Account
	if err := json.Unmarshal(data, &accounts); err != nil {
		return nil, fmt.Errorf("parsing accounts.json: %w", err)
	}
	if accounts == nil {
		accounts = []Account{}
	}
	return accounts, nil
}

// Save atomically persists the account list to accounts.json.
func Save(accounts []Account) error {
	path, err := Path()
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return fmt.Errorf("creating %s: %w", filepath.Dir(path), err)
	}
	data, err := json.MarshalIndent(models.NonNil(accounts), "", "  ")
	if err != nil {
		return fmt.Errorf("serializing accounts: %w", err)
	}
	return settings.WriteFileAtomic(path, data)
}

// Find returns the account with the given id.
func Find(accountID string) (Account, error) {
	accounts, err := Load()
	if err != nil {
		return Account{}, err
	}
	for _, a := range accounts {
		if a.ID == accountID {
			return a, nil
		}
	}
	return Account{}, errors.New("account not found")
}

// Add appends an account to the on-disk list and persists it.
func Add(account Account) error {
	accounts, err := Load()
	if err != nil {
		return err
	}
	return Save(append(accounts, account))
}

// Remove deletes an account from the on-disk list, returning it if present.
func Remove(accountID string) (*Account, error) {
	accounts, err := Load()
	if err != nil {
		return nil, err
	}
	for i, a := range accounts {
		if a.ID == accountID {
			removed := a
			accounts = append(accounts[:i], accounts[i+1:]...)
			if err := Save(accounts); err != nil {
				return nil, err
			}
			return &removed, nil
		}
	}
	return nil, nil
}

// --- keyring helpers ---------------------------------------------------------

// ErrNoSecret reports a keyring entry that does not exist (as opposed to a
// keyring that is locked or unavailable). The OAuth layer classifies it as
// "sign in again", never as a transient failure.
var ErrNoSecret = errors.New("no such keyring entry")

func imapPasswordKey(accountID string) string      { return "imap-password:" + accountID }
func oauthRefreshTokenKey(accountID string) string { return "oauth-refresh-token:" + accountID }

// keyringGet/Set/Delete are swappable so tests never touch the real Secret
// Service.
var (
	keyringGet    = keyring.Get
	keyringSet    = keyring.Set
	keyringDelete = keyring.Delete
)

func getSecret(key string, what string) (string, error) {
	v, err := keyringGet(KeyringService, key)
	if errors.Is(err, keyring.ErrNotFound) {
		return "", fmt.Errorf("reading %s from keyring: %w", what, ErrNoSecret)
	}
	if err != nil {
		return "", fmt.Errorf("reading %s from keyring: %w", what, err)
	}
	return v, nil
}

// SetImapPassword stores an IMAP password for the account.
func SetImapPassword(accountID, password string) error {
	if err := keyringSet(KeyringService, imapPasswordKey(accountID), password); err != nil {
		return fmt.Errorf("storing IMAP password in keyring: %w", err)
	}
	return nil
}

// GetImapPassword fetches the IMAP password for the account.
//
// Development builds only: if COSMIC_MAIL_TEST_IMAP_PASSWORD is set, it is
// returned without touching the keyring, so E2E tests can run against a
// fixture IMAP server on a headless machine with no Secret Service. Compiled
// out of production builds; the keyring path is unchanged for real accounts.
func GetImapPassword(accountID string) (string, error) {
	if buildinfo.Debug {
		if v, ok := os.LookupEnv("COSMIC_MAIL_TEST_IMAP_PASSWORD"); ok {
			return v, nil
		}
	}
	return getSecret(imapPasswordKey(accountID), "IMAP password")
}

// SetOAuthRefreshToken stores the Gmail OAuth refresh token for the account.
func SetOAuthRefreshToken(accountID, token string) error {
	if err := keyringSet(KeyringService, oauthRefreshTokenKey(accountID), token); err != nil {
		return fmt.Errorf("storing OAuth refresh token in keyring: %w", err)
	}
	return nil
}

// GetOAuthRefreshToken fetches the Gmail OAuth refresh token for the account.
func GetOAuthRefreshToken(accountID string) (string, error) {
	return getSecret(oauthRefreshTokenKey(accountID), "OAuth refresh token")
}

// DeleteSecrets removes the account's secrets, best effort (missing entries
// and deletion errors are ignored).
func DeleteSecrets(accountID string, kind models.AccountKind) {
	key := imapPasswordKey(accountID)
	if kind == models.AccountKindGmail {
		key = oauthRefreshTokenKey(accountID)
	}
	_ = keyringDelete(KeyringService, key)
}
